File: 4007.widest-possible-fence/solution.go
func maximumWidth(planks []int) int {
	freq := make(map[int]int)
	for _, p := range planks {
		freq[p]++
	}

	uniquePlanks := make([]int, 0, len(freq))
	for p := range freq {
		uniquePlanks = append(uniquePlanks, p)
	}

	widthForHeight := make(map[int]int)

	for _, p := range uniquePlanks {
		widthForHeight[p] += freq[p]
	}

	for i, p1 := range uniquePlanks {
		if freq[p1] >= 2 {
			widthForHeight[p1*2] += freq[p1] / 2
		}

		for j := i + 1; j < len(uniquePlanks); j++ {
			p2 := uniquePlanks[j]
			sum := p1 + p2
			pairs := min(freq[p1], freq[p2])
			widthForHeight[sum] += pairs
		}
	}

	maxWidth := 0
	for _, width := range widthForHeight {
		maxWidth = max(width, maxWidth)
	}

	return maxWidth
}

File: 4007.widest-possible-fence/solution.kt
class Solution {
    fun maximumWidth(planks: IntArray): Int {
        val frequencies = planks.asIterable().groupingBy { it }.eachCount()
        val widthForHeight = frequencies.toMutableMap()
        val uniquePlanks = frequencies.keys.toList()

        uniquePlanks.forEachIndexed { i, p1 ->
            val frequency1 = frequencies.getOrDefault(p1, 0)
            if (frequency1 >= 2) {
                val frequency2 = widthForHeight.getOrDefault(2 * p1, 0)
                widthForHeight[2 * p1] = frequency2 + frequency1 / 2
            }
            for (j in i + 1 until uniquePlanks.size) {
                val p2 = uniquePlanks[j]
                val sum = p1 + p2
                val frequency2 = widthForHeight.getOrDefault(sum, 0)
                widthForHeight[sum] = frequency2 + minOf(frequency1, frequencies.getValue(p2))
            }
        }

        return widthForHeight.values.max()
    }
}